pkg/helm/watcher/controller/helm_watcher.go

```diff
@@ -0,0 +1,103 @@
+package controller
+
+import (
+	"context"
+
+	sourcev1 "github.com/fluxcd/source-controller/api/v1beta1"
+	"github.com/go-logr/logr"
+	"github.com/helm/helm/pkg/chartutil"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/weaveworks/weave-gitops/pkg/helm"
+	"github.com/weaveworks/weave-gitops/pkg/helm/watcher/cache"
+)
+
+// HelmWatcherReconciler runs the `reconcile` loop for the watcher.
+type HelmWatcherReconciler struct {
+	client.Client
+	Cache       cache.Cache
+	RepoManager helm.HelmRepoManager
+}
+
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories,verbs=get;list;watch
+// +kubebuilder:rbac:groups=helm.watcher,resources=helmrepositories/status,verbs=get
+
+// Reconcile is either called when there is a new HelmRepository or, when there is an update to a HelmRepository.
+// Because the watcher watches all helmrepositories, it will update data for all of them.
+func (r *HelmWatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logr.FromContextOrDiscard(ctx)
+
+	// get source object
+	var repository sourcev1.HelmRepository
+	if err := r.Get(ctx, req.NamespacedName, &repository); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// handle deleted event
+	if repository.DeletionTimestamp != nil {
```

Comment:
Hmpf. This doesn't work for things deleted by hand which don't have the curtesy of adding a deletion timestamp. This might be a problem. I can't definitively detect if something is deleted or not.

Reply:
Hmm.. I might be able to do it in the Get actually. Assume that it couldn't find it, because it was deleted and that's why the recon loop was triggered. That's good enough :)

Reply:
Hum.... upon delete the repository isn't given, meaning, it's empty. so I can't remove the cache for it.

Sorry @aclevername, deleting is not something I can support from the reconciler. And now I remember what finalisers are there for. :D But we don't have that. :)

Reply:
Can you say more about this, what does "deleted by hand" mean here?

Reply:
`kubectl delete`

Reply:
I'm not a 100% sure if flux, when removing something, sets a delete timestamp either. Do you know something, Kevin?

Reply:
Actually, I can just test that. :)

Reply:
Ummm...`kubectl delete` should set the timestamp https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-deletion

Reply:
If it does, the reconcile does not get it. I did a kubectl delete and it didn't have a timestamp.

Reply:
No, I know what's going on. The reconcile needs to get the object in order to see anything about it. But by the time the reconcilre is called the object is no longer there.

```go

	// get source object
	var repository sourcev1.HelmRepository
	if err := r.Get(ctx, req.NamespacedName, &repository); err != nil {
		return ctrl.Result{}, client.IgnoreNotFound(err)
	}

```

This runs and returns, because the object is not found.

Reply:
This is why there are finalisers. You can do your cleanup in a finaliser.

Reply:
https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#finalizers

Which I'm sure you know about. :)

Reply:
Maybe have a look at https://github.com/fluxcd/source-controller/blob/main/controllers/helmrepository_controller.go#L94-L107

Reply:
So this registers a finaliser with:

```
// +kubebuilder:rbac:groups=source.toolkit.fluxcd.io,resources=helmrepositories/finalizers,verbs=get;create;update;patch;delete
```

And somewhere else I can't find. :) I can do that too, but I'm not sure I should.

Reply:
Ok. Gonna add a finaliser. 👍